Sync with upstream

greenc-FNAL opened the pull request:
This time for _sure_.

greenc-FNAL committed:
PackageBase should not define builder legacy attributes (#33942)

* Add a regression test for 33928

* PackageBase should not set `(build|install)_time_test_callbacks`

* Fix audits by preserving the current semantic

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

greenc-FNAL committed:
Revert "Merge pull request #66 from chissg/bugfix/PackageBase-test-callbacks"

This reverts commit 9c020da21df641256b2f799073abc7f714c0f3c3, reversing
changes made to 74a8c307b9d474d58ba843d57526598b7837f427.

haampie committed:
uninstall: fix accidental cubic complexity (#34005)

* uninstall: fix accidental cubic complexity

Currently spack uninstall runs in worst case cubic time complexity
thanks to traversal during traversal during traversal while collecting
the specs to be uninstalled.

Also brings down the number of error messages printed to something
linear in the amount of matching specs instead of quadratic.

scottwittenburg committed:
ci: restore ability to reproduce gitlab job failures (#33953)

climbfuji committed:
ecflow: add v5.8.3, update with changes from JCSDA-EMC fork (#34038)

greenc-FNAL committed:
libjpeg-turbo: build_system settings should be comprehensive

bvanessen committed:
Fixed the rdma-core package to find its external library (#33798)

alalazo committed:
Speed-up a few unit-tests (#34044)

* test_suite.py: speed up slow test by using mock packages

* Don't resolve the sha during unit-tests

* Skip long-running test that fails, instead of executing it

greenc-FNAL committed:
libjpeg-turbo: make build_system settings comprehensive (#34046)

haampie committed:
Spec traversal: add option for topological ordering (#33817)

Spec traversals can now specify a topological ordering. A topologically-
ordered traversal with input specs X1, X2... will

* include all of X1, X2... and their children
* be ordered such that a given node is guaranteed to appear before any
  of its children in the traversal

Other notes:

* Input specs can be children of other input specs (this is useful if
  a user specifies a set of specs to uninstall: some of those specs
  might be children of others)
* `direction="parents"` will produce a reversed topological order
  (children always come before parents).
* `cover="edges"` will generate a list of edges L such that (a) input
  edges will always appear before output edges and (b) if you create
  a list with the destination of each edge in L the result is
  topologically ordered

aweits committed:
qiskit: updates (#33877)

skosukhin committed:
py-fprettify: new version 0.3.7 (#34040)

haampie committed:
gnuconfig: bump with 2022-09-17 (#34035)

genric committed:
py-kubernetes: add version 25.3.0 (#33915)

johnwparent committed:
Windows: add registry query and SDK/WDK packages (#33021)

* Add a WindowsRegistryView class, which can query for existing
  package installations on Windows. This is particularly important
  because some Windows packages (including those added here)
  do not allow two simultaneous installs, and this can be
  queried in order to provide a clear error message.
* Consolidate external path detection logic for Windows into
  WindowsKitExternalPaths and WindowsCompilerExternalPaths objects.
* Add external-only packages win-sdk and wgl
* Add win-wdk (including external detection) which depends on
  win-sdk
* Replace prior msmpi implementation with a source-based install
  (depends on win-wdk). This install can control the install
  destination (unlike the binary installation).
* Update MSVC compiler to choose vcvars based on win-sdk dependency
* Provide "msbuild" module-level variable to packages during build
* When creating symlinks on Windows, need to explicitly specify when
  a symlink target is a directory
* executables_in_path no-longer defaults to using PATH (this is
  now expected to be taken care of by the caller)

becker33 committed:
Warn about removal of deprecated format strings (#33829)

Co-authored-by: becker33 <becker33@users.noreply.github.com>

alalazo committed:
Revert "Warn about removal of deprecated format strings (#33829)" (#34056)

This reverts commit 7f9af8d4a0bfbb1577e5ac9982624d8d0cb9b9ca.

Commit:
eckit: add v1.20.2, v1.16.3 (#33200)

Co-authored-by: Bernhard Kaindl <contact@bernhard.kaindl.dev>

iarspider committed:
cpu-features: Fix appending -DBUILD_SHARED_LIBS (#34055)

Also:
* Use the release tarball for v0.7.0 to fix spack warning

Co-authored-by: Bernhard Kaindl <43588962+bernhardkaindl@users.noreply.github.com>

bernhardkaindl committed:
berkeleygw: fix build (no change to attribute spec.compiler_flags) (#34019)

haampie committed:
spack uninstall: use topo order (#34053)

adamjstewart committed:
Python: drop EOL versions (#33898)

This PR removes [end of life](https://endoflife.date/python) versions of Python from Spack. Specifically, this includes all versions of Python older than 3.7.

See https://github.com/spack/spack/discussions/31824 for rationale. Deprecated in #32615. And #28003.

For anyone using software that relies on Python 2, you have a few options:

* Upgrade the software to support Python 3. The `3to2` tool may get you most of the way there, although more complex libraries may need manual tweaking.
* Add Python 2 as an [external package](https://spack.readthedocs.io/en/latest/build_settings.html#external-packages). Many Python libraries do not support Python 2, but you may be able to add older versions that did once upon a time.
* Use Spack 0.19. Spack 0.19 is the last release to officially support Python 3.6 and older
* Create and maintain your own [custom repository](https://spack.readthedocs.io/en/latest/repositories.html). Basically, you would need a package for Python 2 and any other Python 2-specific libraries you need.

michaelkuhn committed:
cmd/checksum: allow adding new versions to package (#24532)

This adds super-lazy maintainer mode to `spack checksum`: Instead of
only printing the new checksums to the terminal, `-a` and
`--add-to-package` will add the new checksums to the `package.py` file
and open it in the editor afterwards for final checks.

umar456 committed:
span-lite: Add new versions (#34072)

giordano committed:
ccache: add new versions (#34067)

tueda committed:
form: new version 4.3.0 (#34078)

KIwabuchi committed:
Metall package: add v0.22, v0.23, and v0.23.1 (#34073)

hainest committed:
dyninst: add v12.2.1 (#34050)

qwertos committed:
New package: py-rasterstats (#34070)

* Fixed dependencies for rasterstats

* Fixed flake8 errors

* Fix flake8 error

* Cleans up package desc., adds build dependency on setuptools.

* Fixes flake8 error

Co-authored-by: Bailey Brown <bobits@rit.edu>

mdorier committed:
[mochi-margo] fixed dependency to Argobots (#34082)

bernhardkaindl committed:
boost: At least with older Xcode, boost can't build with lzma (#34075)

Reference: https://lists.boost.org/Archives/boost/2019/11/247380.php
As reported at the end of #33998 and this link, liblzma on older Xcode on
MacOSX 10 misses _lzma_cputhreads, so boost's can't use liblzma on those.

alexrobomind committed:
capnproto: update to v0.10.2 (#34063)

Co-authored-by: Alexander Knieps <a.knieps@fz-juelich.de>

haampie committed:
make/ninja: use the right number of jobs (#34057)

adamjstewart committed:
Build System docs: consistent headers (#34047)

haampie committed:
libxcrypt: add v4.33 (#34069)

alecbcs committed:
Add conflicts statements to flux-core to limit builds to linux based platforms (#34068)

cgmb committed:
rocm: add minimum versions for amdgpu_targets (#34030)

glaweh committed:
berkelygw: use mpi variant for scalapack  (#33948)

The package.py assumed "+mpi" in many places, without checking for the variant.
This problem went undetected, as a hard dependency on scalapack pulled an mpi
implementation into the dependency chain (this is also fixed).

Also, the +mpi variant is used select between serial and parallel mode:

It has to enable MPI and ScaLAPACK: They are inter-dependent. Compile
fails because of checks for the other if the other is not enabled.

Co-authored-by: Bernhard Kaindl <contact@bernhard.kaindl.dev>

bernhardkaindl committed:
openloops: add check for added Fortran compiler (#34014)

iarspider committed:
Add checksum for py-parsimonious 0.10.0 (#34079)

iarspider committed:
Add checksum for py-cffi 1.15.1 (#34081)

adamjstewart committed:
libelf: fix build on macOS arm64 (#34036)

qwertos committed:
New package: py-imagecodecs (#34098)

* [libjpeg-turbo] Added version 2.1.3

* [imagecodecs] Added jpeg deependency commented outconglicting libraries

* [WIP]

* [py-imagecodecs] modifying setup.py to work with spack install locations

* [py-imagecodecs] Removed comments and unneeded dependencies

* [py-imagecodecs] removed some comments and fixed up some flake8 complaints

* [py-imagecodecs] flake8

* [py-imagecodecs] fixed import

* [@spackbot] updating style on behalf of qwertos

Co-authored-by: James A Zilberman <jazrc@rit.edu>
Co-authored-by: qwertos <qwertos@users.noreply.github.com>

haampie committed:
Revert "gitlab: Add shared PR mirror to places pipelines look for binaries. (#33746)" (#34087)

This reverts commit 5c4137baf19b9e271a2f13e886d6b875aab067dd.

kwryankrattiger committed:
E4S: Conservatively add ecp-data-vis-sdk (#33621)

* E4S: Conservatively add ecp-data-vis-sdk

* Remove ascent from CUDA SDK stack to stop hanging on Dray

* Adios2: Newer FindPython uses Python_EXECUTABLE

haampie committed:
`url_exists` related improvements (#34095)

For reasons beyond me Python thinks it's a great idea to upgrade HEAD
requests to GET requests when following redirects. So, this PR adds a
better `HTTPRedirectHandler`, and also moves some ad-hoc logic around
for dealing with disabling SSL certs verification.

Also, I'm stumped by the fact that Spack's `url_exists` does not use
HEAD requests at all, so in certain cases Spack awkwardly downloads
something first to see if it can download it, and then downloads it
again because it knows it can download it. So, this PR ensures that both
urllib and botocore use HEAD requests.

Finally, it also removes some things that were there to support currently
unsupported Python versions.

Notice that the HTTP spec [section 10.3.2](https://datatracker.ietf.org/doc/html/rfc2616.html#section-10.3.2) just talks about how to deal
with POST request on redirect (whether to follow or not):

>   If the 301 status code is received in response to a request other
>   than GET or HEAD, the user agent MUST NOT automatically redirect the
>   request unless it can be confirmed by the user, since this might
>   change the conditions under which the request was issued.

>   Note: When automatically redirecting a POST request after
>   receiving a 301 status code, some existing HTTP/1.0 user agents
>   will erroneously change it into a GET request.

Python has a comment about this, they choose to go with the "erroneous change".
But they then mess up the HEAD request while following the redirect, probably
because they were too busy discussing how to deal with POST.

See https://github.com/python/cpython/pull/99731

qwertos committed:
[py-antlr4-python3-runtime] Added versions 4.9.3 and 4.10 (#34102)

* Working updates to py-antlr4-python3-runtime and py-omegaconf

* [py-antlr4-python3-runtime] added version 4.9.3

* [@spackbot] updating style on behalf of qwertos

Co-authored-by: Benjamin Meyers <bsmits@rit.edu>
Co-authored-by: qwertos <qwertos@users.noreply.github.com>

trws committed:
Control Werror by converting to Wno-error (#30882)

Using `-Werror` is good practice for development and testing, but causes us a great
deal of heartburn supporting multiple compiler versions, especially as newer compiler
versions add warnings for released packages.  This PR adds support for suppressing
`-Werror` through spack's compiler wrappers.  There are currently three modes for
the `flags:keep_werror` setting:

* `none`: (default) cancel all `-Werror`, `-Werror=*` and `-Werror-*` flags by
  converting them to `-Wno-error[=]*` flags
* `specific`: preserve explicitly selected warnings as errors, such as
  `-Werror=format-truncation`, but reverse the blanket `-Werror`
* `all`: keeps all `-Werror` flags

These can be set globally in config.yaml, through the config command-line flags, or
overridden by a particular package (some packages use Werror as a proxy for determining
support for other compiler features).  We chose to use this approach because:

1. removing `-Werror` flags entirely broke *many* build systems, especially autoconf
   based ones, because of things like checking `-Werror=feature` and making the
   assumption that if that did not error other flags related to that feature would also work
2. Attempting to preserve `-Werror` in some phases but not others caused similar issues
3. The per-package setting came about because some packages, even with all these
   protections, still use `-Werror` unsafely.  Currently there are roughly 3 such packages
   known.

alalazo committed:
ecflow: polish recipe (#34043)

vvolkl committed:
docs: update info on XCode requirements (#34097)

alkino committed:
HighFive: bump to 2.6.2 (#34090)

adamjstewart committed:
gnuplot: fix build with Apple Clang (#34092)

balay committed:
petsc, slepc: enable parallel builds (#34024)

meyersbs committed:
Add py-urwid@2.1.2 (#34103)

* Add py-urwid@2.1.2

* [@spackbot] updating style on behalf of meyersbs

stephenmsachs committed:
openfoam: Fix openfoam@2012_220610 %intel (add #include <array>) (#34088)

adamjstewart committed:
py-cartopy: older versions don't support newer matplotlib (#34109)

qwertos committed:
New package: py-kt-legacy (#34104)

* first build of keras-tuner with dataset kt-legacy

* [py-kt-legacy] fixed homepage

* [py-kt-legacy] depends on setuptools

* [py-kt-legacy] fixed import

* [@spackbot] updating style on behalf of qwertos

Co-authored-by: Sid Pendelberry <sid@rit.edu>
Co-authored-by: qwertos <qwertos@users.noreply.github.com>

hainest committed:
dyninst: add checksums for all supported versions (#34051)

Thomas-Ulrich committed:
hipsycl: add v0.9.3 (#34052)

brenthuisman committed:
Arbor: Yank v0.5 (#34094)

v0.5 does not build due to a change in setting `arch` introduced in v0.5.2, compatibility with which was not kept in `arbor/package.py`. Since v0.5.2 is compatible with `arbor/package.py`, and is API compatible with v0.5, any users relying on v0.5 can rely on v0.5.2.

iarspider committed:
Add checksum for py-uncertainties 3.1.7 (#34116)

skosukhin committed:
zlib: fix shared libraries when '%nvhpc' (#34039)

haampie committed:
Track locks by (dev, ino); close file handlers between tests (#34122)

adamjstewart committed:
py-pytorch-lightning: add v1.8.3 (#34096)

adamjstewart committed:
py-pandas: add v1.5.2 (#34091)

tmadlener committed:
lcio: Add new version and restrictions on c++ standard (#33997)

snehring committed:
ffmpeg: adding version 5.1.2 (#33758)

* ffmpeg: add version 5.1.2 and switch to conditional variants

Also: py-torchvision: restrict ffmpeg dependency

petertea committed:
Update TotalView versions and website (#33418)

Co-authored-by: Peter Thompson <thompson81@llnl.gov>
Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

victoria-cherkas committed:
metkit, fdb: Add latest versions (#33289)

awvwgk committed:
toml-f: add 0.2.4 and 0.3.1 (#34025)

zzzoom committed:
amdblis: symlink libblis-mt to libblis (#32819)

balay committed:
Revert "`url_exists` related improvements (#34095)" (#34144)

This reverts commit d06fd26c9ac8dd525fc129096188e2ea9fd2d0d7.

The problem is that Bitbucket's API forwards download requests to an S3 bucket using a temporary URL. This URL includes a signature for the request, which embeds the HTTP verb. That means only GET requests are allowed, and HEAD requests would fail verification, leading to 403 erros. The same is observed when using `curl -LI ...`

marcost2 committed:
nvtop: Add 2.0.3, 2.0.4, 3.0.0 and 3.0.1 (#34145)

* And add the option to compile support for Intel GPU's

fangohr committed:
Octopus: branch for Octopus development is now "main" (#34128)

Historically, development of the Octopus code was done on the "develop" branch
on https://gitlab.com/octopus-code/octopus but now development takes place on
"main" (since Q3 2022).

The suggestion in this PR to solve the issue is to keep the spack label
`octopus@develop` as this better indicates this is the development branch on git
than `octopus@main`, but of course to use the `main` branch (there is no choice
here - the `develop` branch is not touched anymore). Sticking to
`octopus@develop` as the version label also keeps backwards compatibility.

luke-dt committed:
minced: add v0.4.2 (#34141)

luke-dt committed:
aragorn: add newer versions and URL (#34140)

wdconinc committed:
prmon: Add missing depends_on py-numpy, py-pandas when +plot (#34123)

berquist committed:
emacs: add variant treesitter for Emacs 29+ (#34134)

luke-dt committed:
py-fastpath: new package (#34142)

* py-fastpath: new package

* Update var/spack/repos/builtin/packages/py-fastpath/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

haampie committed:
neovim: fix deptypes (#34060)

fpruvost committed:
pastix: add new version 6.2.2 (#34066)

usivakum committed:
SingularityCE: Add conmon+squashfs as dependencies  (#33891)

michaelkuhn committed:
rust: add 1.65.0 (#34124)

alecbcs committed:
direnv: add v2.32.2 (#34147)

alecbcs committed:
Add restic v0.14.0 (#34148)

adamjstewart committed:
Add type hints to Prefix class (#34135)

annop-w committed:
armpl-gcc: add post-installation check by running examples (#34086)

haampie committed:
Revert "Revert "gitlab: Add shared PR mirror to places pipelines look for binaries. (#33746)" (#34087)" (#34153)

This reverts commit 63e440651436fdc30993ef58875279e71f4851d1.

alalazo committed:
Use a module-like object to propagate changes in the MRO, when setting build env (#34059)

This fixes an issue introduced in #32340, which changed the semantics of the "module"
object passed to the "setup_dependent_package" callback.

hseara committed:
texinfo: add v7.0 (#34150)

annop-w committed:
acfl: provides blas, lapack, and fftw-api@3 (#34154)

Thomas-Ulrich committed:
easi: update package, rework impalajit (#34032)

manuelakuhn committed:
py-dcm2bids: add new package (#34163)

becker33 committed:
warn about removal of deprecated format strings (#34101)

* warn about removal of deprecated format strings

Co-authored-by: becker33 <becker33@users.noreply.github.com>

iarspider committed:
Add checksum for py-traitlets 5.3.0 (#34127)

manuelakuhn committed:
py-bidskit: new package and dcm2niix: add 1.0.20220720 (#34162)

* py-bidskit: new package and dcm2niix: add 1.0.20220720

* Remove list_url

balay committed:
petsc, py-petsc4py: add 3.18.2 (#34161)

eugeneswalker committed:
e4s ci: add hdf5-vol-async; remove expired comments (#34110)

cwsmith committed:
omegah: new scorec version, fix cuda flags (#34169)

iarspider committed:
Add checksum for py-rsa 4.9 (#34115)

* Add checksum for py-rsa 4.9

* Update package.py

* Update var/spack/repos/builtin/packages/py-rsa/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

greenc-FNAL committed:
Merge branch 'v0.20.0-dev.fermi' into fnal-develop

greenc-FNAL committed:
Merge remote-tracking branch 'upstream/develop' into fnal-develop

meyersbs committed:
New package: py-stui (#34156)

* New package py-stui

* [py-stui] add maintainer

* [@spackbot] updating style on behalf of meyersbs

* [py-stui] fix deps

meyersbs committed:
New package: py-statmorph (#34158)

* New package py-statmorph w/ dependecies. Add py-astropy@5.1

* [@spackbot] updating style on behalf of meyersbs

* [py-statmorph,py-astropy,py-pyerfa] minor fixes

charmoniumQ committed:
Add new version of snakemake (#34041)

* Add new version of snakemake

* Add myself as a maintainer

* py-retry -> py-reretry

* Added snakemake variants for storage systems

* Updated comments

* Responded to Adam's comments

* Fixed spack style

* Add build/run dependency types

fangohr committed:
fix typo in path for sanity check (#34117)

- typo breaks install

vvolkl committed:
genfit: add v02-00-01 (#34159)

psakievich committed:
Add maintainer to Exawind stack and Trilinos (#34174)

* Add maintainer to Nalu-Wind and Trilinos

* Add to trilinos

* Exawind too

* amr-wind too

lpottier committed:
hiredis: updated package definition to use CMake (#33949)

LecrisUT committed:
Fix libxc cflag (#34000)

Using standard c99 should not be specific to intel compilers.

vvolkl committed:
evtgen: add v02.02.00 (#34187)

* evtgen: add v02.02.00
* format

kwryankrattiger committed:
Dav sdk catalyst (#34010)

* SDK: Add Catalyst 1 and 2 support to the SDK

* LibCatalyst: Remove unused python3 variant from package

stephenmsachs committed:
icc@2021.6.0 does not support gcc@12 headers (#34191)

Error message:
```
/shared/spack/opt/spack/linux-amzn2-x86_64_v3/gcc-7.3.1/gcc-12.2.0-4tairupdxg2tg2yhvjdlbs7xbd7wudl3/bin/../include/c++/12.2.0/bits/random.h(104): error: expected a declaration
{ extension using type = unsigned __int128; };
^
```

Co-authored-by: Stephen Sachs <stesachs@amazon.com>

kwryankrattiger committed:
CI: Update Data and Vis SDK Stack (#34009)

* CI: Update Data and Vis SDK Stack

* Update image to match target deployments (E4S)
* Enable all packages
* Test supported variants of ParaView and VisIt

* Sensei: Update Python hint for newer cmake

* Sensei: add Python3 hint

prckent committed:
add 15.0.5, 15.0.6 (#34194)

tgamblin committed:
debug: move "nonexistent config path" message to much higher verbosity level (#34201)

We currently report that searched config paths don't exist at debug level 1, which
clutters the output quite a bit:

```console
> spack -d solve --fresh --show asp hdf5 > hdf5.lp
==> [2022-11-29-14:18:21.035133] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/darwin/concretizer.yaml
==> [2022-11-29-14:18:21.035151] Skipping nonexistent config path /Users/gamblin2/.spack/concretizer.yaml
==> [2022-11-29-14:18:21.035169] Skipping nonexistent config path /Users/gamblin2/.spack/darwin/concretizer.yaml
==> [2022-11-29-14:18:21.035238] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/repos.yaml
==> [2022-11-29-14:18:21.035996] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/defaults/darwin/repos.yaml
==> [2022-11-29-14:18:21.036021] Skipping nonexistent config path /etc/spack/repos.yaml
==> [2022-11-29-14:18:21.036039] Skipping nonexistent config path /etc/spack/darwin/repos.yaml
==> [2022-11-29-14:18:21.036057] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/repos.yaml
==> [2022-11-29-14:18:21.036072] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/darwin/repos.yaml
==> [2022-11-29-14:18:21.036088] Skipping nonexistent config path /Users/gamblin2/.spack/repos.yaml
==> [2022-11-29-14:18:21.036105] Skipping nonexistent config path /Users/gamblin2/.spack/darwin/repos.yaml
==> [2022-11-29-14:18:21.071828] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/config.yaml
==> [2022-11-29-14:18:21.081628] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/defaults/darwin/config.yaml
==> [2022-11-29-14:18:21.081669] Skipping nonexistent config path /etc/spack/config.yaml
==> [2022-11-29-14:18:21.081692] Skipping nonexistent config path /etc/spack/darwin/config.yaml
==> [2022-11-29-14:18:21.081712] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/config.yaml
==> [2022-11-29-14:18:21.081731] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/darwin/config.yaml
==> [2022-11-29-14:18:21.081748] Skipping nonexistent config path /Users/gamblin2/.spack/config.yaml
==> [2022-11-29-14:18:21.081764] Skipping nonexistent config path /Users/gamblin2/.spack/darwin/config.yaml
==> [2022-11-29-14:18:21.134909] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/packages.yaml
==> [2022-11-29-14:18:21.148695] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/darwin/packages.yaml
==> [2022-11-29-14:18:21.152555] Skipping nonexistent config path /etc/spack/packages.yaml
==> [2022-11-29-14:18:21.152582] Skipping nonexistent config path /etc/spack/darwin/packages.yaml
==> [2022-11-29-14:18:21.152601] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/packages.yaml
==> [2022-11-29-14:18:21.152620] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/darwin/packages.yaml
==> [2022-11-29-14:18:21.152637] Skipping nonexistent config path /Users/gamblin2/.spack/packages.yaml
==> [2022-11-29-14:18:21.152654] Skipping nonexistent config path /Users/gamblin2/.spack/darwin/packages.yaml
==> [2022-11-29-14:18:21.853915] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/defaults/compilers.yaml
==> [2022-11-29-14:18:21.853962] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/defaults/darwin/compilers.yaml
==> [2022-11-29-14:18:21.853987] Skipping nonexistent config path /etc/spack/compilers.yaml
==> [2022-11-29-14:18:21.854007] Skipping nonexistent config path /etc/spack/darwin/compilers.yaml
==> [2022-11-29-14:18:21.854025] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/compilers.yaml
==> [2022-11-29-14:18:21.854043] Skipping nonexistent config path /Users/gamblin2/src/spack/etc/spack/darwin/compilers.yaml
==> [2022-11-29-14:18:21.854060] Skipping nonexistent config path /Users/gamblin2/.spack/compilers.yaml
==> [2022-11-29-14:18:21.854093] Reading config from file /Users/gamblin2/.spack/darwin/compilers.yaml
```

It is very rare that I want to know this much information about config search, so I've
moved this to level 3. Now at level 1, we can see much more clearly what configs were
actually found:

```console
> spack -d solve --fresh --show asp hdf5 > hdf5.lp
==> [2022-11-29-14:19:04.035457] Imported solve from built-in commands
==> [2022-11-29-14:19:04.035818] Imported solve from built-in commands
==> [2022-11-29-14:19:04.037626] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/concretizer.yaml
==> [2022-11-29-14:19:04.040033] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/repos.yaml
==> [2022-11-29-14:19:04.080852] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/config.yaml
==> [2022-11-29-14:19:04.133241] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/packages.yaml
==> [2022-11-29-14:19:04.147175] Reading config from file /Users/gamblin2/src/spack/etc/spack/defaults/darwin/packages.yaml
==> [2022-11-29-14:19:05.157896] Reading config from file /Users/gamblin2/.spack/darwin/compilers.yaml
```

You can still get the old messages with `spack -ddd` (to activate debug level 3).

manuelakuhn committed:
py-bidscoin, py-multiecho: add new packages (#34168)

haampie committed:
Single pass text replacement (#34180)

alalazo committed:
Deduplicate code to propagate module changes across MRO (#34157)

s-sajid-ali committed:
PMIx: enable python bindings (#34107)

iarspider committed:
Fix recipe for py-onnx-runtime (#34130)

* Fix recipe

* Update package.py

* Update recipe following review

* Update var/spack/repos/builtin/packages/py-onnx-runtime/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

* remove unused imports

* Update package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

iarspider committed:
Add checksum for py-kiwisolver 1.4.4 (#34121)

eugeneswalker committed:
e4s ci: hpx: set max_cpu_count=512 (#33977)

eugeneswalker committed:
e4s ci: use 2022-12-01 runner images (#34212)

charmoniumQ committed:
Add py-yt 4.x versions (#30418)

* Add py-yt 4.x versions

* Fix spelling

* Add yt dependencies

* Refine cython dependency

* Tweak depends_on for py-yt 4.x

* Apply suggestions from code review

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

* Fix comments from code review

* Fix formatting

* Fix stuff

* Fix constraints

* Update py-yt to 4.1.2

* Updated packages

* Fix py-tomli checksum

* Remove `expand` from `py-tomli/package.py`

* Respond to Adam's comments

* Update checksums

* Update checksusm

* Respond to comments

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

edwardhartnett committed:
new w3emc version (#34219)

* updated version of w3emc package
* fixed sha

adamjstewart committed:
py-segmentation-models-pytorch: add v0.3.1 (#34214)

lpottier committed:
redis-plus-plus: newer version and added TLS support (#34197)

Signed-off-by: Loïc Pottier <pottier1@llnl.gov>

lrknox committed:
Add HDF5 version 1.13.3. (#34165)

* Add HDF5 version 1.13.3.
* Remove maintainers no longer with The HDFGroup.
* Fix indentation.

adamjstewart committed:
py-scipy: hardcode to use blis.pc (#34171)

becker33 committed:
conditional variant values: allow boolean (#33939)

iarspider committed:
gnuplot: make readline optional (#34179)

* gnuplot: make readline optional
* Update package.py

Co-authored-by: iarspider <iarspider@users.noreply.github.com>

alecbcs committed:
Add py-python-lsp-server and dependencies (#34149)

* Add py-python-lsp-server and dependencies

* Update var/spack/repos/builtin/packages/py-python-lsp-server/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

* Relax version range constraints on py-python-lsp-jsonrpc and add missing dep

* Add runtime dependency flag to setuptools dependencies

* Remove unused python@3.6: dependency and move setuptools-scm to build dep only

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

annop-w committed:
acfl: add post-installation check by running examples (#34172)

manuelakuhn committed:
py-twine: add 4.0.1, py-readme-renderer: add 37.3 (#34203)

* py-twine: add 4.0.1

* Remove py-setuptools as run dependency

iarspider committed:
Herwig3: make njet, vbfnlo dependencies optional... (#33941)

* Herwig3: make njet, vbfnlo dependencies optional...
  also drop openloops dependency when building on PowerPC
* Update package.py

manuelakuhn committed:
py-flask: add 2.2.2 and fix dependencies for py-werkzeug and py-markupsafe (#32849)

* py-flask: add 2.2.2, py-werkzeug: add 2.2.2, py-markupsafe: add 2.1.1

* Remove py-dataclasses dependency

tldahlgren committed:
Bugfix: Fetch should not force use of curl to check url existence (#34225)

* Bugfix: Fetch should not force use of curl to check url existence

* Switch type hints from comments to actual hints

iarspider committed:
Add checksum for py-hatchling 1.8.1 (#34260)

manuelakuhn committed:
py-keyrings-alt: add 4.2.0 (#34262)

* py-keyrings-alt: add 4.2.0

* Add missing py-jaraco-classes dependency

marcmengel reviewed (approved):
Looks like a good list of changes from upstream.

tgamblin committed:
concretizer: use only `attr()` for Spec attributes (#31202)

All Spec attributes are now represented as `attr(attribute_name, ... args ...)`, e.g.
`attr(node, "hdf5")` instead of `node("hdf5")`, as we *have* to maintain the `attr()`
form anyway, and it simplifies the encoding to just maintain one form of the Spec
information.

Background
----------

In #20644, we unified the way conditionals are done in the concretizer, but this
introduced a nasty aspect to the encoding: we have to maintain everything we want in
general conditions in two forms: `predicate(...)` and `attr("predicate", ...)`. For
example, here's the start of the table of spec attributes we had to maintain:

```prolog
node(Package)                      :- attr("node", Package).
virtual_node(Virtual)              :- attr("virtual_node", Virtual).
hash(Package, Hash)                :- attr("hash", Package, Hash).
version(Package, Version)          :- attr("version", Package, Version).
...
```

```prolog
attr("node", Package)              :- node(Package).
attr("virtual_node", Virtual)      :- virtual_node(Virtual).
attr("hash", Package, Hash)        :- hash(Package, Hash).
attr("version", Package, Version)  :- version(Package, Version).
...
```

This adds cognitive load to understanding how the concretizer works, as you have to
understand the equivalence between the two forms of spec attributes. It also makes the
general condition logic in #20644 hard to explain, and it's easy to forget to add a new
equivalence to this list when adding new spec attributes (at least two people have been
bitten by this).

Solution
--------

- [x] remove the equivalence list from `concretize.lp`
- [x] simplify `spec_clauses()`, `condition()`, and other functions in `asp.py` that need
      to deal with `Spec` attributes.
- [x] Convert all old-form spec attributes in `concretize.lp` to the `attr()` form
- [x] Simplify `display.lp`, where we also had to maintain a list of spec attributes. Now
      we only need to show `attr/2`, `attr/3`, and `attr/4`.
- [x] Simplify model extraction logic in `asp.py`.

Performance
-----------

This seems to result in a smaller grounded problem (as there are no longer duplicated
`attr("foo", ...)` / `foo(...)` predicates in the program), but it also adds a slight
performance overhead vs. develop. Ultimately, simplifying the encoding will be a win,
particularly for improving error messages.

Notes
-----

This will simplify future node refactors in `concretize.lp` (e.g., not identifying nodes
by package name, which we need for separate build dependencies).

I'm still not entirely used to reading `attr()` notation, but I thnk it's ultimately
clearer than what we did before. We need more uniform naming, and it's now clear what is
part of a solution. We should probably continue making the encoding of `concretize.lp`
simpler and more self-explanatory. It may make sense to rename `attr` to something like
`node_attr` and to simplify the names of node attributes. It also might make sense to do
something similar for other types of predicates in `concretize.lp`.

tldahlgren committed:
Docs: Minor change 'several'->'over a dozen' (#34274)

hyoklee committed:
feat(Hermes): update version to 0.9.0-beta. (#34243)

* feat(Hermes): update version to 0.9.0-beta.
   This is the latest release.
* feat(Hermes): fix checksum.
  Credit: @tldahlgren

manuelakuhn committed:
py-neurokit2: add 0.2.2 (#34267)

sethrj committed:
ROOT: add math/gsl conflict and change version-dependent features to conditional variants (#34244)

* ROOT: add GSL/math dependency
* ROOT: use conditional variants instead of conflicts

becker33 committed:
patch command: add concretizer args (#34282)

* patch command: add concretizer args
* tab completion

manuelakuhn committed:
py-nilearn: fix dependency version (#34284)

aweits committed:
py-alphafold: update to 2.2.4, update dependencies (#33876)

* py-alphafold: update to 2.2.4, update dependencies

* style

snehring committed:
libvips: adding version 8.13.3 (#34228)

Commit:
add mochi-margo 0.11.1 point release (#34271)

drbenmorgan committed:
vecgeom: add new 1.2.1 version (#34240)

* vecgeom: add new 1.2.1 version
* vecgeom: introduce conflict between gcc/cuda

Recent tests of vecgeom in Spack environments have shown that the build
with +cuda fails with GCC >= 11.3 and CUDA < 11.7 with error

...lib/gcc/x86_64-pc-linux-gnu/11.3.0/include/serializeintrin.h(41):
error: identifier "__builtin_ia32_serialize" is undefined

1 error detected in the compilation of
".../VecGeom/source/BVHManager.cu".

Other GCC/CUDA combinations appear o.k.

Avoid this error in spack, and document it for users, with a conflict
directive to express the restriction.

sethrj committed:
doxygen: add build-tools tag (#34249)

* doxygen: add build-tool tag
   This allows it to be included automatically as an external. No one links against doxygen so this should be ok.
* doxygen: add self as maintainer

enrico-usai committed:
aws-parallelcluster: Add v2.11.9 (#34270)

wspear committed:
Use correct method for selecting compiler names. (#34175)

glennpj committed:
py-tensorflow: patch for cuBLAS error (#34279)

adamjstewart committed:
py-torchmetrics: add v0.11.0 (#34220)

tgamblin committed:
typing: move from comment annotations to Python 3.6 annotations (#34305)

We've stopped supporting Python 2, and contributors are noticing that our CI no longer
allows Python 2.7 comment type hints. They end up having to adapt them, but this adds
extra unrelated work to PRs.

- [x] Move to 3.6 type hints across the entire code base

manuelakuhn committed:
py-mne: add 1.2.2, 1.2.3 and dependency packages (#34295)

* py-mne: add 1.2.2 and depencendy packages

* py-mne: add 1.2.3

* Remove unnecessary when statement

tgamblin committed:
web: remove checks for SSL verification support (#34307)

We no longer support Python <3.6, so we don't need to check whether Python supports SSL
verification in `spack.util.web`.

- [x] Remove a bunch of logic we needed to appease Python 2

aurianer committed:
Add conflicts between gcc@12.2.0 and rocblas@5.2.1:5.2.3 (#34217)

* Add conflicts with gcc@12.2.0

* Add more links for reference

adamjstewart committed:
py-torch: fix patching on ppc64le (#34283)

luke-dt committed:
mmseqs2: new package (#34189)

HELICS-bot committed:
helics: add v3.3.2 (#34297)

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

glennpj committed:
wannier90: gfortran-10 support and libs property (#34278)

glennpj committed:
freeglut: add dependency on libxxf86vm (#34293)

glennpj committed:
gurobi: add v10.0.0, v9.5.2 (#34291)

jeanlucf22 committed:
Thermo4PFM: new package (#34287)

sxs-bot committed:
spectre: add v2022.12.02 (#34277)

Co-authored-by: sxs-bot <sxs-bot@users.noreply.github.com>

manuelakuhn committed:
py-llvmlite: add 0.39.1 (#34318)

tgamblin committed:
vermin: remove all novm comments from code (#34308)

All the vermin annotations we were using were for optional features introduced in early
Python 3 versions. We no longer need any of them, as we only support Python 3.6+. If we
start optionally using features from newer Pythons than 3.6, we will need more vermin
annotations.

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

greenc-FNAL committed:
binutils: resolve confusing not-quite-duplication of methods

manuelakuhn committed:
py-nbclassic: add 0.4.8 and new package py-notebook-shim (#34320)

* py-nbclassic: add 0.4.8 and new package py-notebook-shim

* Add missing dependencies

andriish committed:
Add the very first version of cernlib package (#33911)

* Add the very first version of cernlib
* Update package.py
* Update package.py

Co-authored-by: Andrii Verbytskyi <andriish@pcatlas18.mpp.mpg.de>

tgamblin committed:
clingo: add version 5.6.2 (#34317)

See release notes at https://github.com/potassco/clingo/releases/tag/v5.6.2

downloadico committed:
Add packages related to the LWA software stack (#34112)

* epsic: add epsic package to spack

* psrcat: add psrcat to spack

* psrchive: add psarchive to spack

* tempo: add tempo package to spack

hyoklee committed:
feat(Hermes)!: add yaml-cpp dependency (#34330)

The 0.9.0-beta requires yaml-cpp for parsing the configuration file format in YAML.

P.S. I'm using https://www.conventionalcommits.org/en/v1.0.0/#specification for this commit message.

rbberger committed:
LAMMPS: Add new versions (#32522)

* LAMMPS: Add version 20220803 and 20220623.1

* LAMMPS: Add 20220915, 20221103, and 20220623.2

hseara committed:
Added plumed version 2.8.1 including gromacs compatibility (#34268)

* Added plumed version 2.8.1 including gromacs compatibility
* Corrected ~mpi and +mpi variants in new depends
* Fixed regression logic plumed+gromacs@2020.6 support

srekolam committed:
Drop support for older rocm releases - 4.5.0 till 5.0.2 (#34264)

* initial changes for rocm recipes
* drop support for older releases
* drop support for older rocm releases - add more recipes
* drop support for older releases
* address style issues
* address style error
* fix errors
* address review comments

manuelakuhn committed:
py-num2words: add 0.5.12 (#34315)

manuelakuhn committed:
py-nodeenv: add 1.7.0 (#34314)

lpottier committed:
redis: newer version and added TLS support (#34230)

Signed-off-by: Loïc Pottier <pottier1@llnl.gov>
Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

greenc-FNAL committed:
Style fix

greenc-FNAL committed:
Merge remote-tracking branch 'upstream/develop' into chg-job-test

Conflicts:
	lib/spack/spack/solver/asp.py
	lib/spack/spack/solver/concretize.lp
	lib/spack/spack/solver/display.lp
	var/spack/repos/builtin/packages/py-onnx-runtime/package.py

greenc-FNAL commented:
upstream-sync/fnal-develop
